### DISCUSSION TITLE
[Feature Request] Proper permission setting for /etc/v2ray/config.json

### DISCUSSION THREAD

**Discussion by BeautyyuYanli** [v2fly/v2ray-core#2287] [Ideas]
The config file contains all the info (and secrets) of the tunnels, however it is not properly protected. By default, the permission setting of the dir `/etc` doesn't forbiden read access to the files in it. It can be dangerous as all the applications (including those produced in ~somewhere~, e.g. linuxqq) can read users' v2ray config.

My proposal is to add a new user (i.e. `v2ray`) to run the program. The permission setting for `config.json` should be like:
```
-rw-rw---- 1 v2ray v2ray 1024  1月 01 00:00 config.json
```
and `v2ray.service`:
```
[Service]
User=v2ray
```

relate to #428

**Comment by yin1999** [comment]
We can also use [Credentials](https://systemd.io/CREDENTIALS/) with DynamicUser:

```ini
[Service]
DynamicUser=true
LoadCredential=config.json:/etc/v2ray/config.json
ExecStart=/usr/bin/v2ray -config ${CREDENTIALS_DIRECTORY}/config.json
```

**Comment by mydogshitgold** [comment]
IMO you should try avoid running untrustworthy programs on devices with sensitive data. And if you have to, run them in isolated environments like sandboxes or containers, and setup MAC like SeLinux or AppArmor.

Merely changing file permissions won't protect you against such programs, the adversaries could deduce that you are using V2Ray and potentially your server configuration from command hash table, shell history, system caches and browser data.

**Reply by BeautyyuYanli** [reply]
I do use sandbox, but it is still necessary to set proper permissions I think.

**Comment by bogart108** [comment]
msg warn "开始安装..."
    [[ $is_core_ver ]] && msg warn "${is_core_name} 版本: ${yellow}$is_core_ver${none}"
    [[ $proxy ]] && msg warn "使用代理: ${yellow}$proxy${none}"
    # create tmpdir
    mkdir -p $tmpdir
    # if is_core_file, copy file
    [[ $is_core_file ]] && {
        cp -f $is_core_file $is_core_ok
        msg warn "${yellow}${is_core_name} 文件使用 > $is_core_file${none}"
    }
    # local dir install sh script
    [[ $local_install ]] && {
        >$is_sh_ok
        msg warn "${yellow}本地获取安装脚本 > $PWD ${none}"
    }

    timedatectl set-ntp true &>/dev/null
    [[ $? != 0 ]] && {
        msg warn "${yellow}\e[4m提醒!!! 无法设置自动同步时间, 可能会影响使用 VMess 协议.${none}"
    }

    # install dependent pkg
    install_pkg $is_pkg &

    # jq
    if [[ $(type -P jq) ]]; then
        >$is_jq_ok
    else
        jq_not_found=1
    fi
    # if wget installed. download core, sh, jq, get ip
    [[ $is_wget ]] && {
        [[ ! $is_core_file ]] && download core &
        [[ ! $local_install ]] && download sh &
        [[ $jq_not_found ]] && download jq &
        get_ip
    }

    # waiting for background tasks is done
    wait

    # check background tasks status
    check_status

    # test $is_core_file
    if [[ $is_core_file ]]; then
        unzip -qo $is_core_ok -d $tmpdir/testzip &>/dev/null
        [[ $? != 0 ]] && {
            msg err "${is_core_name} 文件无法通过测试."
            exit_and_del_tmpdir
        }
        for i in ${is_core} geoip.dat geosite.dat; do
            [[ ! -f $tmpdir/testzip/$i ]] && is_file_err=1 && break
        done
        [[ $is_file_err ]] && {
            msg err "${is_core_name} 文件无法通过测试."
            exit_and_del_tmpdir
        }
    fi

    # get server ip.
    [[ ! $ip ]] && {
        msg err "获取服务器 IP 失败."
        exit_and_del_tmpdir
    }

    # create sh dir...
    mkdir -p $is_sh_dir

    # copy sh file or unzip sh zip file.
    if [[ $local_install ]]; then
        cp -rf $PWD/* $is_sh_dir
    else
        unzip -qo $is_sh_ok -d $is_sh_dir
    fi

    # create core bin dir
    mkdir -p $is_core_dir/bin
    # copy core file or unzip core zip file
    if [[ $is_core_file ]]; then
        cp -rf $tmpdir/testzip/* $is_core_dir/bin
    else
        unzip -qo $is_core_ok -d $is_core_dir/bin
    fi

    # add alias
    echo "alias $is_core=$is_sh_bin" >>/root/.bashrc

    # core command
    ln -sf $is_sh_dir/$is_core.sh $is_sh_bin

    # jq
    [[ $jq_not_found ]] && mv -f $is_jq_ok /usr/bin/jq

    # chmod
    chmod +x $is_core_bin $is_sh_bin /usr/bin/jq

    # create log dir
    mkdir -p $is_log_dir

    # show a tips msg
    msg ok "生成配置文件..."

    # create systemd service
    load systemd.sh
    is_new_install=1
    install_service $is_core &>/dev/null

    # create condf dir
    mkdir -p $is_conf_dir

    load core.sh
    # create a tcp config
    add tcp
    # remove tmp dir and exit.
    exit_and_del_tmpdir ok
}

# start.
main $@
